lib/chewy/search/parameters.rb

```diff
@@ -0,0 +1,75 @@
+require 'chewy/search/parameters/query'
+require 'chewy/search/parameters/post_filter'
+require 'chewy/search/parameters/limit'
+require 'chewy/search/parameters/offset'
+require 'chewy/search/parameters/order'
+require 'chewy/search/parameters/track_scores'
+require 'chewy/search/parameters/request_cache'
+require 'chewy/search/parameters/explain'
+require 'chewy/search/parameters/version'
+require 'chewy/search/parameters/profile'
+require 'chewy/search/parameters/search_type'
+require 'chewy/search/parameters/preference'
+require 'chewy/search/parameters/terminate_after'
+
+module Chewy
+  module Search
+    class Parameters
+      def self.storages
+        @storages ||= Hash.new do |hash, name|
+          hash[name] = "Chewy::Search::Parameters::#{name.to_s.camelize}".constantize
+        end
+      end
+
+      attr_accessor :storages
+      delegate :[], :[]=, to: :storages
+
+      def initialize(initial = {})
+        @storages = Hash.new do |hash, name|
+          hash[name] = self.class.storages[name].new
+        end
+        initial = initial.map do |name, value|
+          storage_class = self.class.storages[name]
+          storage = value.is_a?(storage_class) ? value : storage_class.new(value)
+          [name, storage]
+        end.to_h
```

Comment:
Using `each_with_object`, it would look like:

```ruby
initial.each_with_object({}) do |(name, value), result|
  storage_class = self.class.storages[name]
  storage = value.is_a?(storage_class) ? value : storage_class.new(value)
  result[name] = storage
end
```